### PR TITLE
Implementing rewards for agents

### LINKED ISSUE #31: Implement target neighbours for each agent
Need a way to tell an agent that they need to be a certain distance away from another agent. Based on the paper by D. Gadjov and Pavel this will be used to compute the second part of the reward function

### PR DESCRIPTION
**Pull request by pandyah5**
This PR addresses the following and fixes #31 :

- [x] Adds functionality to set their own target neighbours
- [x] Initializes all agents such that they have atmost 4 target neighbours one unit away in left, right, top and bottom direction
- [x] The user can specify the agent grid dimensions from config
- [x] Built the reward function to account for global arrangement and local arrangement fixing #32
    - [x] Global: Average distance of all agents from target
    - [x] Local: Deviation from the local ideal arrangement based on adjacent neighbours 

What needs to be done:
- [ ] Implement high negative rewards for collisions and out-of-boundary moves
- [ ] Rigorous unit testing of the reward functions (and all other functions too)

**Note: The PR should be squash merged**

### CODE REVIEW

File: src/nepiada/env/nepiada.py [hunk 1]
```diff
@@ -94,6 +95,16 @@ def render(self):
             # Temporary to print grid for debug purposes until we have a better way to render.
             self.world.grid.render_grid()
             self.world.graph.render_graph(comm=False)
+
+            # This is a debugging code
+            # TODO: Remove later
```
Comment (pandyah5) [inline src/nepiada/env/nepiada.py:100]: I have kept this here, until I get a unit test up

Comment (Thanoshan) [reply src/nepiada/env/nepiada.py:100]: Maybe just keep it commented out for the version we merge

File: src/nepiada/env/nepiada.py [hunk 2]
```diff
@@ -267,22 +279,67 @@ def move_drones(self, actions):
...
+                neighbour = self.world.agents[neighbour_name]
+                neighbour_x = neighbour.p_pos[0]
+                neighbour_y = neighbour.p_pos[1]
+                ideal_x = ideal_distance[0]
+                ideal_y = ideal_distance[1]
+
+                deviation_from_arrangement += np.sqrt((neighbour_x - agent_x - ideal_x)**2 + (neighbour_y - agent_y - ideal_y)**2)
+
+            # Compute the agent's net reward
+            rewards[agent_name] = (Config.global_reward_weight * global_arrangement_reward) + (Config.local_reward_weight * deviation_from_arrangement)
```
Comment (Thanoshan) [inline src/nepiada/env/nepiada.py:343]: Correct me if i'm wrong, but the local_reward will go to 0 as the agents get to its ideal position right?

File: src/nepiada/env/nepiada.py [hunk 3]
```diff
@@ -267,22 +279,67 @@ def move_drones(self, actions):
...
-            action = actions[agent_id]
+        for agent_name in reversed(collided_drones):
+            agent = self.world.agents[agent_name]
+            action = actions[agent_name]
             status = self.world.grid.move_drone(agent, action)
             if status == -2:
                 # Drone collided with another drone
                 pass
 
+    def get_rewards(self):
```
Comment (Thanoshan) [inline src/nepiada/env/nepiada.py:301]: Should the rewards be based on the previous position of the agent vs the current position of the agent? i.e the reward is proportional to the action that the agent has just taken?

**Review by Thanoshan — COMMENTED** [review]
Overall LGTM, some questions about get_rewards.

File: src/nepiada/env/nepiada.py (hunk 2 above)
Comment (pandyah5) [reply src/nepiada/env/nepiada.py:343]: Good catch, I'll need to add a negative sign for the rewards

File: src/nepiada/env/nepiada.py (hunk 1 above)
Comment (pandyah5) [reply src/nepiada/env/nepiada.py:100]: Done!

**Review by Thanoshan — APPROVED** [review]
LGTM. Thanks Hetav!

**Pull request merged**